implement a language with a blob type (`L1`)

zerbina opened the pull request:
## Summary

The `L1` language extends the lower-level `L0` language with an
arbitrarily-sized binary data type (`Blob` type).

## Details

Since languages are currently designed bottom-to-top (i.e., lower-level
to higher-level), the higher-level language extends the lower-level
language, rather than the other way around (i.e., how it is described
in the design document).

The L1 -> L0 lowering is intended to focus on stack layout and
allocation; the `Blob` is a generalization of all aggregate data-types,
of which the exact in-memory layout is irrelevant at this stage.

The current implementation of the lowering focuses only on correctness,
with the chosen stack layout being as simple as it gets.

### Misc

Fix a bug with the output stream draining in `tester`, which led to
the output returned from `exec` sometimes being cut off (and thus test
output comparisons to fail).

saem reviewed (approved):
LGTM

All my suggestions are soft.

zerbina commented on passes/builders.nim:
```diff
@@ -0,0 +1,40 @@
+func initBuilder*[T](): Builder[T] =
+  Builder[T](parent: -1)
+
+template open*[T](bu: var Builder[T], k: T, body: untyped) =
```
The usage is meant to be, for example:
```nim
bu.open Asgn:
  bu.add Node(kind: Local, ...)
  bu.open Copy:
    ...
```
so having `add` part of the name *seemed* off, but `open` doesn't capture the intent all that well either. The `MirBuilder` API in NimSkull uses `subTree` for the routine performing the same thing; maybe that's a good alternative?

saem replied:
Yup, `subTree` works well.

zerbina commented:
I've addressed the comments and implemented the lowering pass, runner, and tests.

The pass really is quite simple; the next higher-level language will either be a language with `Record` and `Array` types (plus associated operations), or a language where integer and float locals support having their address taken.

zerbina commented:
I'm somewhat at a loss as to what is causing the test failure. With the previous commit, the tests still worked [fine](https://github.com/nim-works/phy/actions/runs/10186910841), but after the `debugutils` change, it suddenly doesn't, even though the `pass0/runner` doesn't even import the debug utils.

I suspect there's something wrong with the tester.

saem commented:
> I suspect there's something wrong with the tester.

Either that or some weird line ending business? I read over the code a few times, not seeing what might be the issue, nothing stands out.

zerbina commented:
@saem: That seems to have fixed it.

If the tester program is moved into background right after `readData` in `exec` returns, and the runner program terminates (while producing more output) before the tester instance is moved back to the foreground, the `while` loop exits, and the remaining output is thus not read from the stream. A `readAll` after the `while` loop makes sure the remaining output is read.
